src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs

```diff
@@ -285,7 +285,6 @@ private ServiceCallSite TryCreateOpenGeneric(ServiceDescriptor descriptor, Type
                     }
```

Comment:
nit: maybe we can delete its last usage and delete the `using System.Linq` entirely from this file?
```diff
-using System.Linq;
 using System.Reflection;
 using Microsoft.Extensions.Internal;
 
@@ -24,13 +23,9 @@ internal class CallSiteFactory
         public CallSiteFactory(IEnumerable<ServiceDescriptor> descriptors)
         {
             _stackGuard = new StackGuard();
-            _descriptors = descriptors.ToArray();
-            Populate();
-        }
 
-        private void Populate()
-        {
-            foreach (ServiceDescriptor descriptor in _descriptors)
+            int count = 0;
+            foreach (ServiceDescriptor descriptor in descriptors)
             {
                 Type serviceType = descriptor.ServiceType;
                 if (serviceType.IsGenericTypeDefinition)
@@ -41,7 +36,7 @@ private void Populate()
                     {
                         throw new ArgumentException(
                             SR.Format(SR.OpenGenericServiceRequiresOpenGenericImplementation, descriptor.ServiceType),
-                            "descriptors");
+                            nameof(descriptors));
                     }
 
                     if (implementationType.IsAbstract || implementationType.IsInterface)
@@ -67,7 +62,12 @@ private void Populate()
                 Type cacheKey = descriptor.ServiceType;
                 _descriptorLookup.TryGetValue(cacheKey, out ServiceDescriptorCacheItem cacheItem);
                 _descriptorLookup[cacheKey] = cacheItem.Add(descriptor);
+                ++count;
             }
+
+            _descriptors = new ServiceDescriptor[count];
+            count = 0;
+            foreach (ServiceDescriptor descriptor in descriptors) _descriptors[count++] = descriptor;
         }
```

Reply:
That could be worse.  It'll allocate two enumerators and won't benefit from any optimizations possible based on the concrete type of the enumerable, like using CopyTo if it's a collection. In general, ToArray is a fine thing to use if an array is actually needed.

Reply:
It seems we could store _descriptors in a generic list without using array here `_descriptors = new List<ServiceDescriptor>(descriptors)`

Reply:
What does that help?

Reply:
To get rid of the linq dependency entirely while we are at it. :)

Reply:
FWIW, it's _very_ unlikely this would eliminate System.Linq from the application.  Also, without fully investigating all call paths that can lead to this code, I can't be sure that wouldn't regress; for example, if an input was `someArray.Select(...)`, `ToArray` on that will be able to understand the size and more efficiently copy than can List's ctor.